### PR TITLE
Adding APIs for private link

### PR DESCRIPTION
**Pull request by juniwang**
###  Changes
- Adding APIs for SignalR PrivateEndpointConnection
- Adding new field `upstream`

For both changes, backward compatibilities are took into account. If they are `null`, whether it's from old SDK or newer version of SDK, we won't change anything. Only if it's not `null`, we read their children properties and handle it correspondingly.

Validations documented at [Swagger Validation tools](https://github.com/Azure/adx-documentation-pr/wiki/Swagger-Validation-Tools) all passed.

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Comment by akning-ms** [comment]
@juniwang there is  4 validations failure to fix. you should fix avocado first then modal validation & others

**Comment by juniwang** [comment]
@akning-ms Thanks for following up. Validation errors fixed.

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 1]
```diff
@@ -185,6 +185,208 @@
         }
       }
     },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/SignalR/{resourceName}/privateEndpointConnections/{privateEndpointConnectionName}": {
```
Comment (ravbhatnagar) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: resource type names should be camelcased. so > signalR

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 2]
```diff
@@ -185,6 +185,208 @@
...
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Success. The response indicates the update to private endpoint connection is successful.",
+            "schema": {
+              "$ref": "#/definitions/PrivateEndpointConnection"
+            }
+          },
+          "202": {
```
Comment (ravbhatnagar) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: 201 + provisioning State + azure async operation header is the recommended pattern for long running operations.
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creating-or-updating-resources-asynchronously

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 3]
```diff
@@ -1086,6 +1303,287 @@
...
+          "x-ms-enum": {
+            "name": "PrivateLinkServiceConnectionStatus",
+            "modelAsString": true
+          }
+        },
+        "description": {
+          "description": "The reason for approval/rejection of the connection.",
+          "type": "string"
+        },
+        "actionsRequired": {
```
Comment (ravbhatnagar) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: is this an enum or free text?

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 4]
```diff
@@ -185,6 +185,208 @@
...
+        },
+        "x-ms-long-running-operation": true,
+        "x-ms-examples": {
+          "SignalRPrivateEndpointConnections_Delete": {
+            "$ref": "./examples/SignalRPrivateEndpointConnections_Delete.json"
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.SignalRService/SignalR/{resourceName}/privateLinkResources": {
```
Comment (ravbhatnagar) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: is there a point GET on privateLInkResources.

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 5]
```diff
@@ -1086,6 +1303,287 @@
...
+            "$ref": "#/definitions/PrivateLinkResource"
+          }
+        },
+        "nextLink": {
+          "description": "The URL the client should use to fetch the next page (per server side paging).\r\nIt's null for now, added for future use.",
+          "type": "string"
+        }
+      }
+    },
+    "PrivateLinkResource": {
```
Comment (ravbhatnagar) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: does not comply with ARM resource schema. id, name and type are top level read only properties and all other properties go inside the properties envelope.

**Comment by ravbhatnagar** [comment]
@juniwang - few comments

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json (hunk 3 above)
Comment (ArchangelSDY) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: It's free text.

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json (hunk 4 above)
Comment (ArchangelSDY) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: No. The onboarding spec from NRP only requires listing.

See the similar implementation from storage team: https://docs.microsoft.com/en-us/rest/api/storagerp/privatelinkresources/listbystorageaccount

**Comment by juniwang** [comment]
Hi @ravbhatnagar Could you please review this PR again?

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json (hunk 1 above)
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: Updated

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json (hunk 2 above)
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json]: Updated

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json (hunk 5 above)
Comment (juniwang) [reply specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1593]: Updated

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 6]
```diff
@@ -861,7 +1083,7 @@
           "readOnly": true
         },
         "type": {
-          "description": "The type of the service - e.g. \"Microsoft.SignalRService/SignalR\"",
+          "description": "The type of the resource - e.g. \"Microsoft.SignalRService/SignalR\", \"Microsoft.SignalRService/SignalR/PrivateEndpointConnections\"",
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1086]: >Microsoft.SignalRService/SignalR/PrivateEndpointConnections\" [](start = 98, length = 61)

privateEndpointConnections does not belong here as an example on this definition.  This definition describes a tracked resource specific to your service.  PrivateEndpointConnections are proxy

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 7]
```diff
@@ -1059,7 +1297,7 @@
           }
         },
         "value": {
-          "description": "Value of the feature flag. See Azure SignalR service document https://docs.microsoft.com/azure/azure-signalr/ for allowed values.",
+          "description": "Value of the feature flag. See Azure SignalR service document https://docs.microsoft.com/en-us/azure/azure-signalr/ for allowed values.",
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1300]: >en-us/ [](start = 115, length = 6)

don't include locale in documentation links

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 8]
```diff
@@ -1086,6 +1324,307 @@
...
+          "type": "string"
+        }
+      }
+    },
+    "NetworkACL": {
+      "description": "Network ACL",
+      "type": "object",
+      "properties": {
+        "allow": {
+          "description": "Allowed request types. The value can be: ClientConnection, ServerConnection, RESTAPI.",
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1399]: >ClientConnection, ServerConnection, RESTAPI [](start = 67, length = 43)

if this is value constrained it should be represented by an array of string enums

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 9]
```diff
@@ -1086,6 +1324,307 @@
...
+          "type": "string",
+          "readOnly": true
+        },
+        "name": {
+          "description": "The name of the resource.",
+          "type": "string",
+          "readOnly": true
+        },
+        "type": {
+          "description": "The type of the resource - e.g. \"Microsoft.SignalRService/SignalR\", \"Microsoft.SignalRService/SignalR/PrivateEndpointConnections\"",
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1466]: >\"Microsoft.SignalRService/SignalR\" [](start = 58, length = 36)

Please update the type description so it is valid for this definition

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 10]
```diff
@@ -1086,6 +1324,307 @@
...
+          "type": "string",
+          "readOnly": true
+        }
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1600]: id, name, and type are inherited from the ProxyResource->Resource reference.  You don't need to redeclare them here

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/examples/SignalRPrivateEndpointConnections_Create.json [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            "description": null,
+            "actionsRequired": "None"
+          }
+        },
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/myResourceGroup/providers/Microsoft.SignalRService/SignalR/mySignalRService/privateEndpointConnections/mySignalRService.1fa229cd-bf3f-47f0-8c49-afb36723997e",
+        "name": "mySignalRService.1fa229cd-bf3f-47f0-8c49-afb36723997e",
+        "type": "Microsoft.SignalRService/SignalR/privateEndpointConnections"
+      },
+      "headers": {
+        "Location": "https://management.azure.com/subscriptions/subid/providers/Microsoft.SignalRService/...pathToOperationResult..."
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/examples/SignalRPrivateEndpointConnections_Create.json:58]: >"Location" [](start = 8, length = 10)

Async PUT should use the azure-asyncoperation header if it needs additional operation info, not location.  [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put)

File: specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json [hunk 12]
```diff
@@ -1035,6 +1264,14 @@
...
           "description": "Cross-Origin Resource Sharing (CORS) settings."
+        },
+        "upstream": {
+          "$ref": "#/definitions/ServerlessUpstreamSettings",
+          "description": "Upstream settings when the Azure SignalR is in server-less mode."
+        },
+        "networkACLs": {
+          "$ref": "#/definitions/SignalRNetworkACLs",
+          "description": "Network ACLs"
         }
```
Comment (pilor) [inline specification/signalr/resource-manager/Microsoft.SignalRService/stable/2018-10-01/signalr.json:1275]: this needs to be a new api-version.  Handling it via "patch" semantics with nulls is not a good approach as that is not expected by consumers of your API (a PUT should replace all properties).  It makes what-if analysis on template deployments much more complex, will break users with custom Azure Policies in place that require certain properties, etc...

**Review by pilor — CHANGES_REQUESTED** [review]
:clock1:

**Comment by juniwang** [comment]
Thanks for review, we will have a internal discussion first about the api-version

**Comment by NullMDR** [comment]
@juniwang Suggest to refer to https://github.com/Azure/azure-rest-api-specs/pull/8935 and try to use types in the common folder. Thanks!

**Comment by juniwang** [comment]
Close this PR since we have another one: https://github.com/Azure/azure-rest-api-specs/pull/9377

**Pull request closed without merging**